src/pudl/transform/eia.py

```diff
@@ -1149,133 +1140,171 @@ def fix_balancing_authority_codes_with_state(
-            + Package.get_etl_group_tables("eia923")
-        )
     },
     config_schema={
         "debug": Field(
             bool,
             default_value=False,
             description=(
-                "If True, informational columns will be added into "
-                "boiler_generator_assn"
+                "If True, debugging columns will be added to boiler_generator_assn."
             ),
         ),
     },
     required_resource_keys={"dataset_settings"},
+    io_manager_key="pudl_sqlite_io_manager",
 )
-def eia_transform(context, **eia_transformed_dfs):
-    """Creates DataFrames for EIA Entity tables and modifies EIA tables.
-
-    This function coordinates two main actions: generating the entity tables
-    via ``harvesting()`` and generating the boiler generator associations via
-    ``_boiler_generator_assn()``.
-
-    There is also some removal of tables that are no longer needed after the
-    entity harvesting is finished.
-
-    Args:
-        eia_transformed_dfs (dict): a dictionary of table names (keys) and
-            transformed dataframes (values).
-
-    Returns:
-        Dataframes of entity tables transformed EIA tables.
-    """
+def boiler_generator_assn_eia860(context, **eia_transformed_dfs) -> pd.DataFrame:
+    """Create boiler generator association table."""
+    eia_settings = context.resources.dataset_settings.eia
     # Do some final cleanup and assign appropriate types:
     eia_transformed_dfs = {
         name: convert_cols_dtypes(df, data_source="eia")
         for name, df in eia_transformed_dfs.items()
     }
-
-    # Remove the clean_ prefix from the table names.
+    # Remove the clean_ prefix from the table names... temporary compatibility
     eia_transformed_dfs = {
         table_name.replace("clean_", ""): df
         for table_name, df in eia_transformed_dfs.items()
     }
+    return _boiler_generator_assn(
+        eia_transformed_dfs,
+        eia923_years=eia_settings.eia923.years,
+        eia860_years=eia_settings.eia860.years,
+        debug=context.op_config["debug"],
+    )
 
-    eia_settings = context.resources.dataset_settings.eia
-
-    # create the empty entities df to fill up
-    entities_dfs = {}
 
-    # for each of the entities, harvest the static and annual columns.
-    # the order of the entities matter! the
+def harvested_entity_asset_factory(entity: str) -> AssetsDefinition:
+    """Create an asset definition for the harvested entity tables."""
+    harvestable_assets = (
+        "clean_boiler_fuel_eia923",
+        "clean_boiler_generator_assn_eia860",
+        "clean_boilers_eia860",
+        "clean_coalmine_eia923",
+        "clean_fuel_receipts_costs_eia923",
+        "clean_generation_eia923",
+        "clean_generation_fuel_eia923",
+        "clean_generation_fuel_nuclear_eia923",
+        "clean_generators_eia860",
+        "clean_ownership_eia860",
+        "clean_plants_eia860",
+        "clean_utilities_eia860",
```

Comment:
@cmgosnell Can we / should we pare down this list at all? Are all of them really harvested?

Different entities actually depend on different sets of tables (e.g. not every table contains information about generators or boilers).

@bendnorman if we did have different inputs for each of the entity types, what would be the right way to specify that? I guess we could just have a dictionary here in the asset factory and use whichever set of tables was appropriate.

The proliferation of lists of tables makes me a little nervous though.

Reply:
The link between entity and needed input tables could _probably_ be dynamically generated from the columns in the metadata and the id_cols in `ENTITIES`.

The first step of the harvesting process it to look at all of the input tables to see if those tables actually contain the `id_cols` of the entity. We could in theory do that ahead of time and just with the metadata. But it would require any id columns that are natively reported in tables actually be present in the published table. For instance, the `plants_eia860` table has `utility_id_eia` in it. But you could certainly imagine a world where we have a `plants_utility_assn_eia` table and remove the `utility_id_eia` from `plants_eia860`.

I would be reluctant to hard-code the list of input tables -> harvested entity. Mostly bc sometimes we've missed adding columns when we originally ingest data OR we drop columns all willy nilly. and it would be another step to have to remember to definitely go back and add an input table into a list in here. i.e. if we added a column into a table that would make that table harvestable it would quietly not be harvested.

Reply:
hm looking back at probably Ben's original comment about circular imports and stopping us from being able to just load all assets this way: `load_assets_from_modules([pudl.transform.eia860, pudl.transform.eia923])`...

I think that is the way here to avoid all of the naming of assets.

Reply:
Imma move `occurrence_consistency` to helpers and dynamically generate this list.

Seems like we can't easily identify which tables each harvested entity depends on without having metadata stored for the `clean_*` tables, which sounds to me like it only makes sense if we can do it programmatically.

Reply:
Unfortunately there were additional layers of circular import errors beyond `occurrence_consistency`. I poked a little but gave up and reverted.